### PR TITLE
feat: adding enrollment receiver

### PR DESCRIPTION
**Pull request by andrey-canon**
## Description

This will create external certificates when a user enrolls in a course, generally that certificate will have a grade of 0 and then that will be updated by the certificate_publisher

https://edunext.atlassian.net/browse/FUTUREX-488

## Testing instructions

1. Set ENABLE_CERTIFICATE_PUBLISHER to True
2. Save a course enrollment in the admin panel, then check logs. Change the course enrollment mode to get different logs

### CODE REVIEW

File: eox_nelp/signals/receivers.py [hunk 1]
```diff
@@ -112,3 +115,73 @@ def certificate_publisher(certificate, metadata, **kwargs):  # pylint: disable=u
...
+            ),
+            mode=instance.mode,
+            grade=course_grade.percent,
+            current_status='downloadable' if course_grade.passed else 'not-passing',
+            download_url='',
+            name='',
+        )
+
+        create_external_certificate.delay(
+            external_certificate_data=_generate_external_certificate_data(
```
Comment (johanv26) [inline eox_nelp/signals/receivers.py:175]: Here I found a problem if I create an enrollment from scratch.
Maybe the certificate doesn't exist and resolves
Peek 2023-08-31 15-15
 with 500.

File: eox_nelp/signals/tests/test_receivers.py [hunk 2]
```diff
@@ -242,3 +245,177 @@ def test_alternative_mode(self, create_external_certificate_mock, generate_data_
...
+    @patch("eox_nelp.signals.receivers.create_external_certificate")
+    def test_create_call(self, create_external_certificate_mock, generate_data_mock, course_grade_factory_mock):
+        """Test when the enrollment mode is valid and the asynchronous task is called
+
+        Expected behavior:
+            - _generate_external_certificate_data is called with the right parameters.
+            - create_external_certificate is called with the _generate_external_certificate_data output.
+            - Info was logged.
+        """
+        generate_data_mock.return_value = {
```
Comment (johanv26) [inline eox_nelp/signals/tests/test_receivers.py:320]: Also as above, this is mocked so t when this doesn't exist the process turns into an error.

File: eox_nelp/signals/receivers.py [hunk 3]
```diff
@@ -112,3 +115,73 @@ def certificate_publisher(certificate, metadata, **kwargs):  # pylint: disable=u
...
+
+    - ENABLE_CERTIFICATE_PUBLISHER<boolean>: If this is true the receiver will publish the certificate data,
+    default is False.
+    - CERTIFICATE_PUBLISHER_VALID_MODES<list[string]>: List of valid modes, default ['no-id-professional']
+
+    Note: This keeps the same certificate receiver settings since this will create an external certificate at
+    the beginning of the course, then the certificate receiver will update the grade.
+
+    Args:
+        instance<CourseEnrollment>: This an instance of the model CourseEnrollment.
```
[the remark this answers is not in the document]
Comment (andrey-canon) [reply eox_nelp/signals/receivers.py:133]: that names depends on how the signal is emitted

File: eox_nelp/signals/receivers.py (hunk 1 above)
Comment (andrey-canon) [reply eox_nelp/signals/receivers.py:175]: Good catch, the good new is that you found a line that is not necessary

File: eox_nelp/signals/tests/test_utils.py [hunk 4]
```diff
@@ -135,17 +122,13 @@ def test_invalid_group_codes(self, generate_certificate_mock, passing_mock):
 
     @override_settings(EXTERNAL_CERTIFICATES_GROUP_CODES={"course-v1:test+Cx105+2022_T4": "ABC123"})
     @patch("eox_nelp.signals.utils._user_has_passing_grade")
-    @patch("eox_nelp.signals.utils.GeneratedCertificate")
     @data(*WRONG_NATIONAL_IDS)
-    def test_invalid_mational_id(self, wrong_national_id, generate_certificate_mock, passing_mock):
+    def test_invalid_mational_id(self, wrong_national_id, passing_mock):
```
Comment (johanv26) [inline eox_nelp/signals/tests/test_utils.py:126]:
```suggestion
    def test_invalid_national_id(self, wrong_national_id, passing_mock):
```

**Review by johanv26 — COMMENTED** [review]
A little typo to improve in tests but with the change It seems the bug was removed.
 
cut

I think `id` was something good to keep, but no way it could be extracted using the reference_id I think.

**Review by johanv26 — APPROVED** [review]

**Pull request merged**